File: src/nayuki/deflate/GzipDecompress.java
package nayuki.deflate;

import java.io.BufferedInputStream;
import java.io.DataInput;
import java.io.DataInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.Date;
import java.util.zip.CRC32;


public class GzipDecompress {
	
	public static void main(String[] args) {
		String msg = submain(args);
		if (msg != null) {
			System.err.println(msg);
			System.exit(1);
		}
	}
	
	
	private static String submain(String[] args) {
		// Check arguments
		if (args.length != 2)
			return "Usage: java GzipDecompress InputFile OutputFile";
		
		File inFile = new File(args[0]);
		if (!inFile.exists())
			return "Input file does not exist: " + inFile;
		if (inFile.isDirectory())
			return "Input file is a directory: " + inFile;
		
		try {
			// Start reading
			DataInputStream in = new DataInputStream(new BufferedInputStream(new FileInputStream(inFile), 16 * 1024));
			byte[] decomp;
			int crc, size;
			try {
				// Header
				int flags;
				{
					byte[] b = new byte[10];
					in.readFully(b);
					if (b[0] != 0x1F || b[1] != (byte)0x8B)
						return "Invalid GZIP magic number";
					if (b[2] != 8)
						return "Unsupported compression method: " + (b[2] & 0xFF);
					flags = b[3] & 0xFF;
					
					// Reserved flags
					if ((flags & 0xE0) != 0)
						return "Reserved flags are set";
					
					// Modification time
					int mtime = (b[4] & 0xFF) | (b[5] & 0xFF) << 8 | (b[6] & 0xFF) << 16 | b[7] << 24;
					if (mtime != 0)
						System.out.println("Last modified: " + new Date(mtime * 1000L));
					else
						System.out.println("Last modified: N/A");
					
					// Extra flags
					switch (b[8] & 0xFF) {
						case 2:   System.out.println("Extra flags: Maximum compression");  break;
						case 4:   System.out.println("Extra flags: Fastest compression");  break;
						default:  System.out.println("Extra flags: Unknown");  break;
					}
					
					// Operating system
					String os;
					switch (b[9] & 0xFF) {
						case   0:  os = "FAT";             break;
						case   1:  os = "Amiga";           break;
						case   2:  os = "VMS";             break;
						case   3:  os = "Unix";            break;
						case   4:  os = "VM/CMS";          break;
						case   5:  os = "Atari TOS";       break;
						case   6:  os = "HPFS";            break;
						case   7:  os = "Macintosh";       break;
						case   8:  os = "Z-System";        break;
						case   9:  os = "CP/M";            break;
						case  10:  os = "TOPS-20";         break;
						case  11:  os = "NTFS";            break;
						case  12:  os = "QDOS";            break;
						case  13:  os = "Acorn RISCOS";    break;
						case 255:  os = "Unknown";         break;
						default :  os = "Really unknown";  break;
					}
					System.out.println("Operating system: " + os);
				}
				
				// Handle assorted flags
				if ((flags & 0x01) != 0)
					System.out.println("Flag: Text");
				if ((flags & 0x04) != 0) {
					System.out.println("Flag: Extra");
					byte[] b = new byte[2];
					in.readFully(b);
					int len = (b[0] & 0xFF) | (b[1] & 0xFF) << 8;
					in.readFully(new byte[len]);  // Skip extra data
				}
				if ((flags & 0x08) != 0)
					System.out.println("File name: " + readNullTerminatedString(in));
				if ((flags & 0x02) != 0) {
					byte[] b = new byte[2];
					in.readFully(b);
					System.out.printf("Header CRC-16: %04X%n", (b[0] & 0xFF) | (b[1] & 0xFF) << 8);
				}
				if ((flags & 0x10) != 0)
					System.out.println("Comment: " + readNullTerminatedString(in));
				
				// Decompress
				try {
					decomp = Decompressor.decompress(new ByteBitInputStream(in));
				} catch (FormatException e) {
					return "Invalid or corrupt compressed data: " + e.getMessage();
				}
				
				// Footer
				{
					byte[] b = new byte[8];
					in.readFully(b);
					crc  = (b[0] & 0xFF) | (b[1] & 0xFF) << 8 | (b[2] & 0xFF) << 16 | b[3] << 24;
					size = (b[4] & 0xFF) | (b[5] & 0xFF) << 8 | (b[6] & 0xFF) << 16 | b[7] << 24;
				}
			} finally {
				in.close();
			}
			
			// Check
			if (size != decomp.length)
				return String.format("Size mismatch: expected=%d, actual=%d", size, decomp.length);
			if (crc != getCrc32(decomp))
				return String.format("CRC-32 mismatch: expected=%08X, actual=%08X", crc, getCrc32(decomp));
			
			// Write decompressed data to output file
			OutputStream out = new FileOutputStream(args[1]);
			try {
				out.write(decomp);
			} finally {
				out.close();
			}
		} catch (IOException e) {
			return "I/O exception: " + e.getMessage();
		}
		
		return null;
	}
	
	
	private static String readNullTerminatedString(DataInput in) throws IOException {
		StringBuilder sb = new StringBuilder();
		while (true) {
			byte c = in.readByte();
			if (c == 0)
				break;
			else
				sb.append((char)(c & 0xFF));
		}
		return sb.toString();
	}
	
	
	private static int getCrc32(byte[] data) {
		CRC32 crc = new CRC32();
		crc.update(data);
		return (int)crc.getValue();
	}
	
}

File: python/gzip-decompress.py
# 
# Simple DEFLATE decompressor (Python)
# 
# Copyright (c) Project Nayuki
# MIT License. See readme file.
# https://www.nayuki.io/page/simple-deflate-decompressor
# 

import datetime, pathlib, sys, zlib
from typing import Dict, List, Optional
import deflatedecompress


def main(argv: List[str]) -> Optional[str]:
	# Handle command line arguments
	if len(argv) != 3:
		return f"Usage: python {argv[0]} InputFile.gz OutputFile"
	infile = pathlib.Path(argv[1])
	if not infile.exists():
		return f"Input file does not exist: {infile}"
	if infile.is_dir():
		return f"Input file is a directory: {infile}"
	outfile = pathlib.Path(argv[2])
	
	try:
		# Start reading
		with infile.open("rb") as inp:
			
			# Define helper read functions based on 'inp'
			
			def read_byte() -> int:
				b: bytes = inp.read(1)
				if len(b) == 0:
					raise EOFError()
				return b[0]
			
			def read_little_int16() -> int:
				temp: int = read_byte()
				return temp | read_byte() << 8
			
			def read_little_int32() -> int:
				temp: int = read_little_int16()
				return temp | read_little_int16() << 16
			
			def read_null_terminated_string() -> str:
				sb = bytearray()
				while True:
					b: int = read_byte()
					if b == 0:
						return sb.decode("UTF-8")
					sb.append(b)
			
			
			# Header
			if read_byte() != 0x1F:
				return "Invalid GZIP magic number"
			if read_byte() != 0x8B:
				return "Invalid GZIP magic number"
			compmeth: int = read_byte()
			if compmeth != 8:
				return f"Unsupported compression method: {str(compmeth)}"
			flags: int = read_byte()
			
			# Reserved flags
			if flags & 0xE0 != 0:
				return "Reserved flags are set"
			
			# Modification time
			mtime: int = read_little_int32()
			if mtime != 0:
				dt = datetime.datetime.fromtimestamp(mtime, datetime.timezone.utc)
				print(f"Last modified: {dt}")
			else:
				print("Last modified: N/A")
			
			# Extra flags
			extraflags: int = read_byte()
			if extraflags == 2:
				print("Extra flags: Maximum compression")
			elif extraflags == 4:
				print("Extra flags: Fastest compression")
			else:
				print(f"Extra flags: Unknown ({extraflags})")
			
			# Operating system
			OPERATING_SYSTEMS: Dict[int,str] = {
				  0: "FAT",
				  1: "Amiga",
				  2: "VMS",
				  3: "Unix",
				  4: "VM/CMS",
				  5: "Atari TOS",
				  6: "HPFS",
				  7: "Macintosh",
				  8: "Z-System",
				  9: "CP/M",
				 10: "TOPS-20",
				 11: "NTFS",
				 12: "QDOS",
				 13: "Acorn RISCOS",
				255: "Unknown",
			}
			osbyte: int = read_byte()
			osstr: str = OPERATING_SYSTEMS.get(osbyte, "Really unknown")
			print(f"Operating system: {osstr}")
			
			# Handle assorted flags
			if flags & 0x01 != 0:
				print("Flag: Text")
			if flags & 0x04 != 0:
				print("Flag: Extra")
				count: int = read_little_int16()
				while count > 0:  # Skip extra data
					n: int = len(inp.read(count))
					if n == 0:
						raise EOFError()
					count -= n
			if flags & 0x08 != 0:
				print(f"File name: {read_null_terminated_string()}")
			if flags & 0x02 != 0:
				print(f"Header CRC-16: {read_little_int16():04X}")
			if flags & 0x10 != 0:
				print(f"Comment: {read_null_terminated_string()}")
			
			# Decompress
			try:
				bitin = deflatedecompress.BitInputStream(inp)
				decomp: bytes = deflatedecompress.Decompressor.decompress_to_bytes(bitin)
			except ValueError as e:
				return f"Invalid or corrupt compressed data: {e}"
			
			# Footer
			crc : int = read_little_int32()
			size: int = read_little_int32()
		
		# Check decompressed data's length and CRC
		if size != len(decomp) % 2**32:
			return f"Size mismatch: expected={size}, actual={len(decomp)}"
		actualcrc = zlib.crc32(decomp) & 0xFFFFFFFF
		if crc != actualcrc:
			return f"CRC-32 mismatch: expected={crc:08X}, actual={actualcrc:08X}"
		
		# Write decompressed data to output file
		with outfile.open("wb") as out:
			out.write(decomp)
		
	except IOError as e:
		return f"I/O exception: {e}"
	return None  # Success, no error message

	
if __name__ == "__main__":
	errmsg = main(sys.argv)
	if errmsg is not None:
		sys.exit(errmsg)
